src/ppm.rs

```diff
@@ -0,0 +1,473 @@
+    GetRandom(#[from] getrandom::Error),
+}
+
+/// A share of an input or proof for Prio.
+pub enum PrioShare<F: FieldElement> {
+    /// An uncompressed share, typically sent to the leader.
+    Leader(Vec<F>),
+
+    /// A compressed share, typically sent to the helper. The first parameter is a seed for a
+    /// pseudorandom generator
+    Helper(Vec<u8>, usize),
+}
+
+impl<F: FieldElement> TryFrom<PrioShare<F>> for Vec<F> {
+    type Error = PpmError;
+
+    fn try_from(share: PrioShare<F>) -> Result<Self, PpmError> {
+        match share {
+            PrioShare::Leader(data) => Ok(data),
+            PrioShare::Helper(seed, length) => {
+                let mut prng = Prng::from_seed(&seed)?;
+                Ok((0..length).map(|_| prng.next().unwrap()).collect())
+            }
+        }
+    }
+}
+
+/// The message sent by the client to each aggregator. This includes the client's input share and
+/// the initial message of the input-validation protocol.
+pub struct PrioUpload<F: FieldElement> {
+    /// The identity of the intended recipient.
+    pub aggregator_id: u8,
+
+    /// The input share.
+    pub input_share: PrioShare<F>,
+
+    /// The proof share.
+    pub proof_share: PrioShare<F>,
+
+    /// The sum of the joint randomness seed shares sent to the other aggregators.
+    pub joint_rand_seed_hint: Vec<u8>,
+
+    /// The blinding factor, used to derive the aggregator's joint randomness seed share.
+    pub blind: Vec<u8>,
+}
+
+fn assert_safe() {
+    // TODO(cjpatton) Replace this hack with a more consistent API for the Prio ciphersuite.Prng
+    // should be constructed with an implementation of the following trait:
+    //
+    // pub trait PrioCipher {
+    //    // Panics if `seed.len() != StreamCipher::KEY_LEN`.
+    //    fn new_stream_cipher(seed: &[u8]) -> StreamCipher;
+    //
+    //    // Panics if `Hasher::OUT_LEN != StreamCipher::KEY_LEN`.
+    //    fn new_hasher() -> Hasher;
+    // }
+    //
+    // pub trait StreamCipher {
+    //    const KEY_LEN: usize;
+    //
+    //    /// Fill `out` with the next `out.len()` bytes of the stream cipher.
+    //    fn fill(out: &mut [u8]);
+    // }
+    //
+    // pub trait Hasher<const OUT_LEN: usize> {
+    //    const OUT_LEN: usize;
+    //
+    //    /// Appends `input` to the hash input.
+    //    fn update(input: &[u8]);
+    //
+    //    /// Panics if `out.len() != Self::OUT_LEN`.
+    //    fn finalize(out: &mut [u8]);
+    // }
+    if blake3::KEY_LEN != SEED_LENGTH || blake3::OUT_LEN != SEED_LENGTH {
+        panic!("Can't instantiate hash function!");
+    }
+}
+
+/// Run by the client, this generates the sequence of [`PrioUpload`] messages to send to the
+/// aggregators.
+///
+/// # Parameters
+///
+/// * `input` is the input to be secret shared.
+/// * `num_shares` is the number of input shares (i.e., aggregators) to generate.
+pub fn prio_upload<F, V>(input: &V, num_shares: usize) -> Result<Vec<PrioUpload<F>>, PpmError>
```

Comment:
I agree with you that it's difficult to imagine having deployments with more than a single digit number of helpers, let alone more than 255. IIUC this is important because we need to define the length of the aggregator ID that gets hashed into the seed hint. We haven't settled on boundaries for this at the PPM layer yet (which I think should be covered by [#138](https://github.com/abetterinternet/ppm-specification/issues/138), so I think we should settle that matter in PPM and then reflect the decision here.

Reply:
SGTM.